### PR TITLE
Dom/format

### PR DESCRIPTION
**Pull request by domoritz**
Fix https://github.com/vega/vega-lite/issues/1539

### CODE REVIEW

File: src/compile/common.ts [hunk 1]
```diff
@@ -111,24 +110,29 @@ export function formatSignalRef(fieldDef: FieldDef<string>, specifiedFormat: str
...
     }
+
     // TODO: need to make this work correctly for numeric ordinal / nominal type
     return config.numberFormat;
   }
   return undefined;
 }
 
+function formatExpr(field: string, format: string) {
+  return `format(${field}, ${format ? `'${format}'` : null})`;
```
Comment (kanitw) [inline src/compile/common.ts]: So format `''`  is different from `null`?

File: src/compile/common.ts [hunk 2]
```diff
@@ -111,24 +110,29 @@ export function formatSignalRef(fieldDef: FieldDef<string>, specifiedFormat: str
...
+}
+
+export function numberFormatExpr(field: string, specifiedFormat: string, config: Config) {
+  return formatExpr(field, specifiedFormat || config.numberFormat);
```
Comment (kanitw) [inline src/compile/common.ts]: '' || x = x, so this could be wrong?

File: src/compile/axis/parse.ts [hunk 3]
```diff
@@ -120,6 +120,8 @@ function getSpecifiedOrDefaultValue(property: keyof VgAxis, specifiedAxis: Axis,
   switch (property) {
     case 'labels':
       return isGridAxis ? false : specifiedAxis[property];
+    case 'labelOverlap':
+      return rules.labelOverlap(fieldDef, specifiedAxis, channel, isGridAxis);  // TODO: scaleType
```
Comment (kanitw) [inline src/compile/axis/parse.ts]: This doesn't read the  specified `labelOverlap`

File: src/vega.schema.ts [hunk 4]
```diff
@@ -335,6 +335,15 @@ export interface VgAxisBase {
   labelAngle?: number;
 
   /**
+   * Automatic overlap removal.
+   *
+   * `true` or `"parity"`: Reduce overlap by attempting to remove every other label, repeating this process until overlap is removed or only two labels remain. If only two labels remain, the first and last label will be shown. This setting is appropriate for linear, sqrt, and pow settings.
+   * `"greedy"`: Reduce overlap by sequentially scanning the labels, removing any items whose bounds overlap the last visible label. This strategy works better than "parity" for log scales, though is not "ideal" in that it does not actually reason about which ticks are the most useful for interpreting a log scale.
+   * __Default value:__ `true` for x-axes, otherwise `false`
```
Comment (kanitw) [inline src/vega.schema.ts]: The default condition specified does not perfectly match what you implemented

File: src/compile/axis/parse.ts (hunk 3 above)
Comment (kanitw) [inline src/compile/axis/parse.ts]: Also, you can easily getScaleComponent and read the type here to handle overlapping for discrete and log scales.

File: src/compile/common.ts (hunk 1 above)
Comment (domoritz) [inline src/compile/common.ts]: idk, but https://github.com/vega/vega-lite/issues/1539#issuecomment-303504194 suggests that we should use null. It's a bit more explicit.

File: src/compile/common.ts (hunk 2 above)
Comment (domoritz) [inline src/compile/common.ts]: Fixed

File: src/compile/common.ts (hunk 1 above)
Comment (kanitw) [inline src/compile/common.ts]: ok

File: src/compile/axis/parse.ts (hunk 3 above)
Comment (domoritz) [inline src/compile/axis/parse.ts]: Done

File: src/compile/common.ts (hunk 1 above)
Comment (domoritz) [inline src/compile/common.ts]: Okay, I revise. I get an error now. So we need to set the format to ''.

**Comment by domoritz** [comment]
I have no idea what's going wrong with the schema. Will look at it tomorrow.

**Comment by domoritz** [comment]
Needs https://github.com/vega/vega-parser/pull/41

**Pull request merged**